### DISCUSSION TITLE
Broadcast?

### DISCUSSION THREAD

**Discussion by debshaw13** [kartikk221/hyper-express#30] [Q&A] [answered]
Does hyper-express have an equivalent function to broadcast (that is available in ws)? For example, using ws, the following code broadcasts data without any connections:

```
const wss = new WebSocket.Server({ port: 8082 });

wss.broadcast = (data) => {
  wss.clients.forEach((client) => {
    client.send(data);
  });
};
```

If I replace 'WebSocket' above with 'HyperExpress' (after importing 'hyper-express' of course), I see the following error: 

> TypeError: Cannot read property 'forEach' of undefined

Would appreciate any advice as to what I am doing wrong. Thank you for the great package, and apologies if this is a newbie question.

**Comment by kartikk221** [comment, marked as answer]
You could go about this two ways. One method would be where you keep track of all incoming websocket connections and then create your own "broadcast" method which dispatches the data to the clients.
Example:
```javascript
const connections = {};
const uuid_v4 = require('crypto').randomUUID;
const webserver = new HyperExpress.Server();

function ws_broadcast(data){
    Object.keys(connections).forEach((id) => {
        connections[id].send(data);
    });
}

webserver.ws('/connect', (ws) => {
    // Assign some random identifier to the connection
    ws.id = uuid_v4();

    // Store the connection in our connections object
    connections[ws.id] = ws;

    // Remove the connection from our object once it is closed
    ws.on('close', () => {
        delete connections[ws.id];
     });
});
```

The second way is to use the MQTT based pub/sub system which is documented in our documentation.
I would highly recommend you read https://github.com/kartikk221/hyper-express/blob/master/docs/Websocket.md to get more information.

**Reply by debshaw13** [reply]
Thank you for your detailed response! I did not realize the MQTT based pub/sub system is similar to broadcasting. I will re-read the documentation as you have suggested. Thanks again, much appreciated.
